File: macros_post/src/main/scala/russoul/lib/common/math/algebra/Vec.scala
package russoul.lib.common.math.algebra

import russoul.lib.common._
import russoul.lib.macros.Ref
import singleton.ops.{Id, XInt}
import spire.algebra.{Field, VectorSpace}

import scala.collection.TraversableLike
import scala.reflect.ClassTag
import spire.algebra._
import spire.math._
import spire.implicits._
import singleton.ops._
import spire.{NoImplicit, sp}

/**
  * Created by russoul on 11.07.2017.
  */
//size_arg is not a field !
class Vec[@specialized(Int, Float, Double, Long) A : ClassTag, Size <: XInt]private(size_arg: Int) extends Traversable[A]{

  type E = A
  type N = Size

  private val array = new Array[A](size_arg)

  @inline def apply(i: Int): A = array(i)

  def toArray = array.clone()

  override def toString() : String = {
    var str = ""

    for(i <- array) str += i + " "

    if(array.length > 1) str = str.dropRight(1)

    s"Vec[${implicitly[ClassTag[A]].toString()}, ${array.length}]\n$str"
  }

  override def foreach[U](f: (A) => U): Unit = {
    var k = 0
    while (k < array.length){
      f(array(k))
      k += 1
    }
  }


  override def hashCode() = {
    array.hashCode()
  }

  override def equals(obj: scala.Any) = {
    obj match {
      case that : Vec[A, Size] => this.hashCode() == that.hashCode()
      case _ => false
    }
  }




}

object Vec {
  @inline def apply[@specialized(Int, Float, Double, Long) A : ClassTag, Size <: XInt](args: A*): Vec[A, Size] = {
    val result = new Vec[A,Size](args.size)

    var k = 0
    while(k < args.size){
      result.array(k) = args(k)
      k += 1
    }


    result
  }

}

final class VecIsModule[@specialized(Float, Double, Int, Long) F : ClassTag : Ring, Size <: XInt](implicit _size : ValueOf[Size]) extends Module [Vec[F,Size],F] {
  override def scalar: Ring[F] = Ring[F]

  override def timesl(r: F, v: Vec[F, Size]): Vec[F, Size] = {
    val ar = new Array[F](v.size)

    var i = 0
    while (i < ar.size){
      ar(i) = r * v(i)
      i += 1
    }

    Vec[F, Size](ar : _*)
  }

  override def negate(v: Vec[F, Size]): Vec[F, Size] = {
    val ar = new Array[F](v.size)

    var i = 0
    while (i < ar.size){
      ar(i) = -v(i)
      i += 1
    }

    Vec[F, Size](ar : _*)
  }

  override def zero: Vec[F, Size] = {
    val ar = new Array[F](_size.value)

    var i = 0
    while (i < ar.size){
      ar(i) = scalar.zero
      i += 1
    }

    Vec[F, Size](ar : _*)
  }

  override def plus(a: Vec[F, Size], b: Vec[F, Size]): Vec[F, Size] = {
    val ar = new Array[F](a.size)

    var i = 0
    while (i < a.size){
      ar(i) = a(i) + b(i)
      i += 1
    }

    Vec[F, Size](ar : _*)
  }
}

final class VecIsVectorSpace[@specialized(Float, Double) F : ClassTag : Field, Size <: XInt](implicit _size : ValueOf[Size]) extends VectorSpace [Vec[F,Size],F]{
  override def scalar: Field[F] = Field[F]

  override def timesl(r: F, v: Vec[F, Size]): Vec[F, Size] = {
    val ar = new Array[F](v.size)

    var i = 0
    while (i < ar.size){
      ar(i) = r * v(i)
      i += 1
    }

    Vec[F, Size](ar : _*)
  }

  override def negate(v: Vec[F, Size]): Vec[F, Size] = {
    val ar = new Array[F](v.size)

    var i = 0
    while (i < ar.size){
      ar(i) = -v(i)
      i += 1
    }

    Vec[F, Size](ar : _*)
  }

  override def zero: Vec[F, Size] = {
    val ar = new Array[F](_size.value)

    var i = 0
    while (i < ar.size){
      ar(i) = scalar.zero
      i += 1
    }

    Vec[F, Size](ar : _*)
  }

  override def plus(a: Vec[F, Size], b: Vec[F, Size]): Vec[F, Size] = {
    val ar = new Array[F](a.size)

    var i = 0
    while (i < a.size){
      ar(i) = a(i) + b(i)
      i += 1
    }

    Vec[F, Size](ar : _*)
  }
}




final class VecIsNormedVectorSpace[@specialized(Float, Double) F : ClassTag : Field, Size <: XInt](implicit _size : ValueOf[Size], nroot : NRoot[F]) extends NormedVectorSpace [Vec[F,Size],F]{
  override def scalar: Field[F] = Field[F]

  override def timesl(r: F, v: Vec[F, Size]): Vec[F, Size] = {
    val ar = new Array[F](v.size)

    var i = 0
    while (i < ar.size){
      ar(i) = r * v(i)
      i += 1
    }

    Vec[F, Size](ar : _*)
  }

  override def negate(v: Vec[F, Size]): Vec[F, Size] = {
    val ar = new Array[F](v.size)

    var i = 0
    while (i < ar.size){
      ar(i) = -v(i)
      i += 1
    }

    Vec[F, Size](ar : _*)
  }

  override def zero: Vec[F, Size] = {
    val ar = new Array[F](_size.value)

    var i = 0
    while (i < ar.size){
      ar(i) = scalar.zero
      i += 1
    }

    Vec[F, Size](ar : _*)
  }

  override def plus(a: Vec[F, Size], b: Vec[F, Size]): Vec[F, Size] = {
    val ar = new Array[F](a.size)

    var i = 0
    while (i < a.size){
      ar(i) = a(i) + b(i)
      i += 1
    }

    Vec[F, Size](ar : _*)
  }

  override def norm(v: Vec[F, Size]): F = {
    var ret = scalar.zero
    var i = 0
    while (i < v.size){
      ret += v(i)
      i += 1
    }

    ret.sqrt()
  }
}


final class VecIsInnerProductSpace[@specialized(Double, Float) F : ClassTag : Field, Size <: XInt](implicit _size : ValueOf[Size]) extends InnerProductSpace [Vec[F,Size],F]{
  override def scalar: Field[F] = Field[F]

  override def timesl(r: F, v: Vec[F, Size]): Vec[F, Size] = {
    val ar = new Array[F](v.size)

    var i = 0
    while (i < ar.size){
      ar(i) = r * v(i)
      i += 1
    }

    Vec[F, Size](ar : _*)
  }

  override def negate(v: Vec[F, Size]): Vec[F, Size] = {
    val ar = new Array[F](v.size)

    var i = 0
    while (i < ar.size){
      ar(i) = -v(i)
      i += 1
    }

    Vec[F, Size](ar : _*)
  }

  override def zero: Vec[F, Size] = {
    val ar = new Array[F](_size.value)

    var i = 0
    while (i < ar.size){
      ar(i) = scalar.zero
      i += 1
    }

    Vec[F, Size](ar : _*)
  }

  override def plus(a: Vec[F, Size], b: Vec[F, Size]): Vec[F, Size] = {
    val ar = new Array[F](a.size)

    var i = 0
    while (i < a.size){
      ar(i) = a(i) + b(i)
      i += 1
    }

    Vec[F, Size](ar : _*)
  }

  override def dot(a: Vec[F, Size], b: Vec[F, Size]): F = {
    var i = 0
    var r = scalar.zero
    while (i < a.size){
      r += a(i) * b(i)
      i += 1
    }

    r
  }


}

trait VecInstanceM1 {
  type NI0[F,S <: XInt] = NoImplicit[VectorSpace[Vec[F,S], F]]

  implicit def vecModule[@sp(Int,Long,Float,Double) F: ClassTag: Ring, Size <: XInt : ValueOf](implicit no : NI0[F,Size]): VecIsModule[F, Size] =
    new VecIsModule[F,Size]
}

trait VecInstance0 extends VecInstanceM1{
  type NI1[F,S <: XInt] = NoImplicit[InnerProductSpace[Vec[F,S], F]]

  implicit def vecIsVectorSpace[@sp(Float, Double) F : ClassTag : Field, Size <: XInt : ValueOf](implicit no : NI1[F,Size]) : VecIsVectorSpace[F,Size] = {
    new VecIsVectorSpace[F, Size]()
  }
}

trait VecInstance1 extends VecInstance0{
  implicit def vecIsInnerProductSpace[@sp(Float, Double) F : ClassTag : Field, Size <: XInt : ValueOf] : VecIsInnerProductSpace[F,Size] = {
    new VecIsInnerProductSpace[F, Size]()
  }
}


trait VecInstance2 extends VecInstance1{
  implicit def vecIsNormedVectorSpace[@sp(Float, Double) F : ClassTag : Field : NRoot, Size <: XInt : ValueOf] : VecIsNormedVectorSpace[F,Size] = {
    new VecIsNormedVectorSpace[F, Size]()
  }
}

trait VecInstances extends VecInstance2

object VecInstances extends VecInstances

File: src/main/deprecated/Vec.java
package russoul.lib.common.math.algebra.deprecated;

/**
  * Created by russoul on 01.06.2017.
  */

//TODO redundant, do not use this class, use array instead
@immutable class Vec[@specialized(Float,Double,Int) F : ClassTag](private val array:Array[F], var isColumn:Boolean = true)(implicit ev: Field[F]){


  @inline @straight def x: F = array(0)
  @inline @straight def y: F = array(1)
  @inline @straight def z: F = array(2)
  @inline @straight def w: F = array(3)

  @inline def dim():Int = array.size
  @inline def size():Int = array.size

  @inline def foreach[U](f: (F => U)): Unit ={
    array.foreach(f)
  }

  @inline def filter[U](f: (F => Boolean)): Vec[F] ={
    new Vec(array.filter(f), isColumn)
  }

  @inline def min[B >: F]()(implicit ordering : Ordering[B]): F ={
    array.min(ordering)
  }

  @inline def setAsColumn(): Vec[F] = {
    isColumn = true
    this
  }

  @inline def setAsRow(): Vec[F] = {
    isColumn = false
    this
  }

  @inline @straight def apply(index: Int): F = {
    array(index-1)
  }

  @inline def *(scalar: F): Vec[F] = {
    val ar = new Array[F](array.size)
    for(i <- array.indices){
      ar(i) = array(i) * scalar
    }

    new Vec[F](ar)
  }

  @inline def /(scalar: F): Vec[F] = {
    val ar = new Array[F](array.size)
    for(i <- array.indices){
      ar(i) = array(i) / scalar
    }

    new Vec[F](ar)
  }


  @inline @straight def +(that: Vec[F]):Vec[F] = {
    assert(this.size() == that.size())
    val ar = new Array[F](array.size)
    for(i <- array.indices){
      ar(i) = this.array(i) + that.array(i)
    }


    new Vec[F](ar)
  }

  @inline @straight def -(that: Vec[F]):Vec[F] = {
    assert(this.size() == that.size())
    val ar = new Array[F](array.size)
    for(i <- array.indices){
      ar(i) = this.array(i) - that.array(i)
    }

    new Vec[F](ar)
  }

  @inline def ??+(that: Vec[F]): Boolean = {
    this.array.size == that.array.size
  }

  @inline def ??-(that: Vec[F]): Boolean = {
    this.array.size == that.array.size
  }

  @inline def unary_-():Vec[F] = {
    val ar = new Array[F](array.size)
    for(i <- 0 until array.size){
      ar(i) = -this.array(i)
    }

    new Vec[F](ar)
  }

  @inline def ??⨯(that:Mat[F]): Boolean ={
    if(isColumn) that.rows == 1 else that.rows == dim
  }

  @inline @straight def ⨯(that:Mat[F]): Mat[F] ={
    val p = that.columns

    if(isColumn){
      assert(1 == that.rows)
      val ar = new Array[F](dim * p)
      val res = new Mat[F](dim, p, ar)

      for(i <- 1 to dim){
        for(j <- 1 to p){
          res(i,j) = res(i,j) + this(i) * that(1,j)
        }
      }

      res
    }else{
      assert(size() == that.rows)
      val ar = new Array[F](1 * p)
      val res = new Mat[F](1, p, ar)

      for(j <- 1 to p){
        for(k <- 1 to dim) {
          res(1,j) = res(1,j) + this(k) * that(k,j)
        }
      }

      res
    }
  }

}
